internal/speakerlist/speakerlist.go

```diff
@@ -80,36 +83,107 @@ func New(logger gokitlog.Logger, nodeName, bindAddr, bindPort, secret, namespace
 }
 
 // Start starts the needed goroutines
-func (sl *SpeakerList) Start(client *k8s.Client) error {
+func (sl *SpeakerList) Start(client *k8s.Client) {
 	sl.client = client
 
 	if sl.mList == nil {
-		return nil
+		return
 	}
 
 	go sl.watchEvents()
+	go sl.memberlistJoinLoop()
+}
 
-	iplist, err := sl.client.GetPodsIPs(sl.namespace, sl.labels)
-	if err != nil {
-		sl.l.Log("op", "startup", "error", err, "msg", "failed to get PodsIPs")
-		return err
+func (sl *SpeakerList) memberlistJoin() {
+	if sl.mList == nil {
+		return
+	}
+	joinList := sl.joinList()
+	if len(joinList) > 0 {
+		n, err := sl.mList.Join(joinList)
+		sl.l.Log("op", "ml-join", "msg", "Memberlist join", "nb joigned", n, "error ?", err, "join list", joinList)
+	}
+}
+
+func (sl *SpeakerList) joinList() []string {
+	joinList := []string{}
+	sl.Lock()
+	defer sl.Unlock()
+	if len(sl.k8sSpeakers) == 0 {
+		return joinList
+	}
+
+	mlSpeakers := map[string]bool{}
+	for _, n := range sl.mList.Members() {
+		mlSpeakers[n.Name] = true
+	}
+	for name, ip := range sl.k8sSpeakers {
+		if _, ok := mlSpeakers[name]; !ok {
+			joinList = append(joinList, ip)
+		}
+	}
+	return joinList
+}
+
+func (sl *SpeakerList) memberlistJoinLoop() {
+	ticker := time.NewTicker(10 * time.Second)
+	for {
+		select {
+		case <-sl.stopCh:
+			return
+		case <-ticker.C:
+			sl.memberlistJoin()
+		}
 	}
-	n, err := sl.mList.Join(iplist)
-	sl.l.Log("op", "startup", "msg", "Memberlist join", "nb joigned", n, "error ?", err)
+}
 
-	return nil
+// SetSpeakersK8S update the list of speakers from the K8S API point of view
+func (sl *SpeakerList) SetSpeakersK8S(eps *v1.Endpoints) {
+	newmap := map[string]string{}
+	for _, subset := range eps.Subsets {
+		for _, ep := range subset.Addresses {
+			newmap[*ep.NodeName] = ep.IP
+		}
+	}
+	sl.Lock()
+	oldmap := sl.k8sSpeakers
+	sl.Unlock()
+	if !reflect.DeepEqual(oldmap, newmap) {
+		sl.l.Log("op", "setSpeakersK8S", "msg", "New K8S Speakers list set", "old-list", oldmap, "new-list", newmap)
+		sl.Lock()
+		sl.k8sSpeakers = newmap
+		sl.Unlock()
+		sl.memberlistJoin()
+		// needed for split brain and disabled MemberList
+		sl.client.ForceSync()
+	}
 }
 
-// UsableSpeakers return a map of usable nodes
-func (sl *SpeakerList) UsableSpeakers() map[string]bool {
-	if sl.mList != nil {
-		activeNodes := map[string]bool{}
-		for _, n := range sl.mList.Members() {
-			activeNodes[n.Name] = true
+// UsableSpeakers return a map of usable nodes or an error
+func (sl *SpeakerList) UsableSpeakers() (map[string]bool, error) {
+	sl.Lock()
+	defer sl.Unlock()
+	if len(sl.k8sSpeakers) == 0 {
+		return nil, errors.New("k8sSpeakersNotSet")
+	}
+	activeNodes := map[string]bool{}
+	if sl.mList == nil {
+		for name := range sl.k8sSpeakers {
+			activeNodes[name] = true
 		}
-		return activeNodes
+		return activeNodes, nil
+	}
+
+	members := sl.mList.Members()
+	if len(sl.k8sSpeakers) > 1 && float64(len(members)) < float64(len(sl.k8sSpeakers))/2 {
```

Comment:
Can you please elaborate on what will different nodes see (in as k8sSpeakers and other relevant variables) in different scenarios, to make try to catch all possible cases for this `if`? I mean, like node A is totally disconnected (network unplugged) then in sl.k8sSpekears has value X in node A, but in other nodes it is Y, etc. It smells like the tricky part :-)

Reply:
Why this marked as solved? @champtar I'm not sure I follow

Reply:
not trying to handle split brain anymore, so this is not relevant